File: edgequake/sdks/java/src/main/java/io/edgequake/sdk/resources/LineageService.java
package io.edgequake.sdk.resources;

import io.edgequake.sdk.internal.HttpHelper;
import io.edgequake.sdk.models.LineageModels.*;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;

/**
 * Lineage operations: entity lineage, document lineage, chunk detail/lineage,
 * entity provenance, and lineage export.
 *
 * WHY: These endpoints provide full provenance tracking — tracing entities
 * back to source documents, chunks, and line ranges. Required by the
 * improve-lineage mission for all SDKs.
 *
 * @see edgequake/crates/edgequake-api/src/handlers/lineage.rs
 */
public class LineageService {

    private final HttpHelper http;

    public LineageService(HttpHelper http) { this.http = http; }

    // ── Entity Lineage ───────────────────────────────────────────────

    /**
     * Get entity lineage — which documents contributed to an entity.
     * GET /api/v1/lineage/entities/{name}
     */
    public EntityLineageResponse entityLineage(String entityName) {
        return http.get("/api/v1/lineage/entities/" + encode(entityName),
                null, EntityLineageResponse.class);
    }

    // ── Document Lineage ─────────────────────────────────────────────

    /**
     * Get document graph lineage — which entities were extracted from a document.
     * GET /api/v1/lineage/documents/{id}
     */
    public DocumentGraphLineageResponse documentLineage(String documentId) {
        return http.get("/api/v1/lineage/documents/" + encode(documentId),
                null, DocumentGraphLineageResponse.class);
    }

    /**
     * Get full document lineage with metadata.
     * GET /api/v1/documents/{id}/lineage
     */
    public DocumentFullLineageResponse documentFullLineage(String documentId) {
        return http.get("/api/v1/documents/" + encode(documentId) + "/lineage",
                null, DocumentFullLineageResponse.class);
    }

    /**
     * Export document lineage as raw data.
     * GET /api/v1/documents/{id}/lineage/export?format=json
     *
     * WHY: Returns the export payload. Use format="json" or "csv".
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> exportLineage(String documentId, String format) {
        String fmt = (format != null && !format.isEmpty()) ? format : "json";
        return http.get(
                "/api/v1/documents/" + encode(documentId) + "/lineage/export?format=" + fmt,
                null, Map.class);
    }

    // ── Chunk Detail & Lineage ───────────────────────────────────────

    /**
     * Get chunk detail.
     * GET /api/v1/chunks/{id}
     */
    public ChunkDetailResponse chunkDetail(String chunkId) {
        return http.get("/api/v1/chunks/" + encode(chunkId),
                null, ChunkDetailResponse.class);
    }

    /**
     * Get chunk lineage with parent document references.
     * GET /api/v1/chunks/{id}/lineage
     */
    public ChunkLineageResponse chunkLineage(String chunkId) {
        return http.get("/api/v1/chunks/" + encode(chunkId) + "/lineage",
                null, ChunkLineageResponse.class);
    }

    // ── Entity Provenance ────────────────────────────────────────────

    /**
     * Get entity provenance — source documents, chunks, and related entities.
     * GET /api/v1/entities/{id}/provenance
     */
    public EntityProvenanceResponse entityProvenance(String entityId) {
        return http.get("/api/v1/entities/" + encode(entityId) + "/provenance",
                null, EntityProvenanceResponse.class);
    }

    // ── Helpers ──────────────────────────────────────────────────────

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }
}

File: edgequake/sdks/kotlin/src/main/kotlin/io/edgequake/sdk/resources/LineageService.kt
package io.edgequake.sdk.resources

import io.edgequake.sdk.internal.HttpHelper
import io.edgequake.sdk.models.*
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

/**
 * Lineage service — entity and document lineage tracking.
 *
 * WHY: Covers all 7 lineage endpoints from the API surface:
 *   GET /api/v1/lineage/entities/{name}
 *   GET /api/v1/lineage/documents/{id}
 *   GET /api/v1/documents/{id}/lineage
 *   GET /api/v1/documents/{id}/lineage/export?format=
 *   GET /api/v1/chunks/{id}
 *   GET /api/v1/chunks/{id}/lineage
 *   GET /api/v1/entities/{id}/provenance
 *
 * @see edgequake/crates/edgequake-api/src/handlers/lineage.rs
 */
class LineageService(private val http: HttpHelper) {

    /** Get entity lineage — which documents contributed to an entity. */
    fun entityLineage(entityName: String): EntityLineageResponse {
        // WHY: URLEncoder.encode(String, Charset) was added in Java 10, but we use
        // the String overload for broader compatibility with Java 8+.
        val encoded = URLEncoder.encode(entityName, "UTF-8")
        return http.get("/api/v1/lineage/entities/$encoded")
    }

    /** Get document graph lineage — entities and relationships from a document. */
    fun documentLineage(documentId: String): DocumentGraphLineageResponse =
        http.get("/api/v1/lineage/documents/$documentId")

    /** Get full document lineage with metadata. */
    fun documentFullLineage(documentId: String): DocumentFullLineageResponse =
        http.get("/api/v1/documents/$documentId/lineage")

    /** Export lineage data in specified format (json or csv). */
    fun exportLineage(documentId: String, format: String = "json"): Map<String, Any?> =
        http.get("/api/v1/documents/$documentId/lineage/export?format=$format")

    /** Get chunk detail with extracted entities and relationships. */
    fun chunkDetail(chunkId: String): ChunkDetailResponse =
        http.get("/api/v1/chunks/$chunkId")

    /** Get chunk lineage with parent document references. */
    fun chunkLineage(chunkId: String): ChunkLineageResponse =
        http.get("/api/v1/chunks/$chunkId/lineage")

    /** Get entity provenance — full source traceability. */
    fun entityProvenance(entityId: String): EntityProvenanceResponse =
        http.get("/api/v1/entities/$entityId/provenance")
}
